python/cudf/cudf/core/tools/datetimes.py

```diff
@@ -652,3 +658,220 @@ def _isin_datetimelike(
+    (such as years and months). Specifying `freq` with a mixed fixed and
+    non-fixed frequency is currently unsupported. For example:
+
+    >>> cudf.date_range(
+    ...     start='2021-08-23 08:00:00',
+    ...     freq=cudf.DateOffset(months=2, days=5),
+    ...     periods=5)
+    ...
+    NotImplementedError: Mixing fixed and non-fixed frequency offset is
+    unsupported.
+
+    Examples
+    --------
+    >>> cudf.date_range(
+    ...     start='2021-08-23 08:00:00',
+    ...     freq=cudf.DateOffset(years=1, months=2),
+    ...     periods=5)
+    DatetimeIndex(['2021-08-23 08:00:00', '2022-10-23 08:00:00',
+                '2023-12-23 08:00:00', '2025-02-23 08:00:00',
+                '2026-04-23 08:00:00'],
+                dtype='datetime64[ns]')
+
+    """
+    if tz is not None:
+        raise NotImplementedError("tz is currently unsupported.")
+
+    if normalize:
+        raise NotImplementedError(
+            "normalizing timestamps to midnight is unsupported."
+        )
+
+    if closed is not None:
+        raise NotImplementedError("closed is currently unsupported.")
+
+    if [start, end, periods, freq].count(None) > 1:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and freq, exactly "
+            "three must be specified"
+        )
+
+    dtype = np.dtype("<M8[ns]")
+
+    _periods_not_specified = False
+
+    if freq is not None:
+        if isinstance(freq, DateOffset):
+            offset = freq
+        else:
+            raise TypeError("`freq` must be a cudf.DateOffset object.")
+
+        if _check_mixed_freqeuency(offset):
+            raise NotImplementedError(
+                "Mixing fixed and non-fixed frequency offset is unsupported."
+            )
+        offset = offset.to_pandas()
+
+        if start is None:
+            end = cudf.Scalar(end, dtype=dtype)
+            start = cudf.Scalar(
+                pd.Timestamp(end.value) - (periods - 1) * offset, dtype=dtype
+            )
+        elif end is None:
+            start = cudf.Scalar(start, dtype=dtype)
+        elif periods is None:
+            start = cudf.Scalar(start, dtype=dtype)
+            end = cudf.Scalar(end, dtype=dtype)
+            periods = math.ceil(
+                int(end - start) / _offset_to_nanoseconds_lower_bound(offset)
+            )
+            _periods_not_specified = True
+
+            if periods < 0:
+                # end < start, return empty column
+                periods = 0
+            elif periods == 0:
+                # end == start, return exactly 1 timestamp (start)
+                periods = 1
+    else:
+        start = (cudf.Scalar(start, dtype=dtype).value.astype("int64"),)
+        end = (cudf.Scalar(end, dtype=dtype).value.astype("int64"),)
+        arr = cp.linspace(start=start, stop=end, num=periods,)
+        result = cudf.core.column.as_column(arr).astype("datetime64[ns]")
+        return cudf.DatetimeIndex._from_data({None: result})
```

Comment:
I think the `name` kwarg is being dropped here. 

```
>>> cudf.date_range('2011-01-01', '2011-01-10', periods=10, name='cool')
DatetimeIndex(['2011-01-01', '2011-01-02', '2011-01-03', '2011-01-04',
               '2011-01-05', '2011-01-06', '2011-01-07', '2011-01-08',
               '2011-01-09', '2011-01-10'],
              dtype='datetime64[ns]')
>>> type(cudf.date_range('2011-01-01', '2011-01-10', periods=10, name='cool').name)
<class 'NoneType'>
```

I also ran across this scenario when probing this:

```
>>> cudf.date_range('2011-01-01', '2011-01-10', cudf.DateOffset(days=1), name='cool').name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nfs/brmiller/anaconda3/envs/pr_8886/lib/python3.8/site-packages/cudf/core/tools/datetimes.py", line 801, in date_range
    arr = cp.linspace(start=start, stop=end, num=periods,)
  File "/home/nfs/brmiller/anaconda3/envs/pr_8886/lib/python3.8/site-packages/cupy/_creation/ranges.py", line 151, in linspace
    if num < 0:
TypeError: '<' not supported between instances of 'DateOffset' and 'int'
```

Reply:
Thanks for the `name` catch!

The second is likely due to `periods` is specified with a `cudf.DateOffset` object. With this many parameters, it's advised to use keyword arguments instead:
```python
cudf.date_range(start='2011-01-01', end='2011-01-10', freq=cudf.DateOffset(days=1), name='cool').name
'cool'
```